internal/internal_event_handlers.go

```diff
@@ -227,27 +240,25 @@ func (wc *workflowEnvironmentImpl) Complete(result []byte, err error) {
 	wc.completeHandler(result, err)
 }
 
-func (wc *workflowEnvironmentImpl) RequestCancelWorkflow(domainName, workflowID, runID string) error {
-	if domainName == "" {
-		return errors.New("need a valid domain, provided empty")
-	}
-	if workflowID == "" {
-		return errors.New("need a valid workflow ID, provided empty")
-	}
+func (wc *workflowEnvironmentImpl) RequestCancelExternalWorkflow(domainName, workflowID, runID string, callback resultHandler) {
 
 	isChild, decision := wc.decisionsHelper.requestCancelExternalWorkflowExecution(domainName, workflowID, runID)
 	if isChild {
 		// this is for child workflow
 		childWorkflow := decision.getData().(*scheduledChildWorkflow)
 		if childWorkflow.handled {
-			return nil
-		}
-		if childWorkflow.workflowExecution != nil && (decision.isDone() || !childWorkflow.waitForCancellation) {
+			// already have the result of child workflow execution, so the cancellation request is a success (bypass)
+			callback(nil, nil)
+		} else if childWorkflow.workflowExecution != nil && (decision.isDone() || !childWorkflow.waitForCancellation) {
+			// child workflow execution is cancelled, so the cancellation request is a success (bypass)
 			childWorkflow.handle(nil, ErrCanceled)
+			callback(nil, nil)
```

Comment:
same as above, do not call callback until you see the event from server. you might need to save the callback function to the childWorkflow instance.
also, there is an else case where you don't call callback() and the future will block forever.